### PR TITLE
JAMES-3885 Change username - migrate quotas

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/QuotaUsernameChangeTaskStep.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private Mono<Void> setQuotaForNewUser(QuotaRoot quotaRoot, QuotaManager.Quotas quotas) {
+        return setMaxQuota(quotaRoot, quotas)
+            .then(setCurrentQuota(quotaRoot, quotas));
+    }
+
+    private Mono<Void> setMaxQuota(QuotaRoot quotaRoot, QuotaManager.Quotas quotas) {
+        return Mono.zip(setMaxMessagesQuota(quotaRoot, quotas).thenReturn(quotaRoot),
+                setMaxStorageQuota(quotaRoot, quotas).thenReturn(quotaRoot))
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/QuotaUsernameChangeTaskStep.java:81]: `thenReturn` for `Mono.zip`
It for case
Mono.zip(publisher1,publisher2)
if `publisher1` returns empty, publisher2 can not be executed.

File: mailbox/api/src/main/java/org/apache/james/mailbox/quota/MaxQuotaManager.java [hunk 2]
```diff
@@ -53,6 +53,10 @@ public interface MaxQuotaManager {
      */
     void setMaxStorage(QuotaRoot quotaRoot, QuotaSizeLimit maxStorageQuota) throws MailboxException;
 
+    default Publisher<Void> setMaxStorageReactive(QuotaRoot quotaRoot, QuotaSizeLimit maxStorageQuota) {
+        return Mono.fromRunnable(Throwing.runnable(() -> setMaxStorage(quotaRoot, maxStorageQuota)).sneakyThrow());
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply mailbox/api/src/main/java/org/apache/james/mailbox/quota/MaxQuotaManager.java:57]: I did reactive for `CassandraPerUserMaxQuotaManager` 

For another implement, IMO it's unnecessary

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/QuotaUsernameChangeTaskStep.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @Override
+    public int priority() {
+        return 3;
+    }
+
+    @Override
+    public Publisher<Void> changeUsername(Username oldUsername, Username newUsername) {
+        return Mono.fromCallable(() -> userQuotaRootResolver.forUser(oldUsername))
+            .flatMap(quotaRoot -> Mono.from(quotaManager.getQuotasReactive(quotaRoot)))
+            .flatMap(quotas -> setQuotaForNewUser(userQuotaRootResolver.forUser(newUsername), quotas));
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/QuotaUsernameChangeTaskStep.java:71]: Different quotaRoot, one for oldUsername, one for newUsername!

**Comment by Arsnael** [comment]
Thoughts: I get it for max quotas. but regarding the current one is the migration necessary? Shouldn't it be updated automatically with the listeners when we migrate mailboxes and emails? Or I miss something?

**Comment by vttranlina** [comment]
> Thoughts: I get it for max quotas. but regarding the current one is the migration necessary? Shouldn't it be updated automatically with the listeners when we migrate mailboxes and emails? Or I miss something?

I just tested (use `MemoryUsernameChangeIntegrationTest` ), when migrating mailboxes and emails, James doesn't  publish new Add events, 
So, NO event for `ListeningCurrentQuotaUpdater` -> No automatic update current quota

**Comment by Arsnael** [comment]
No you are right, we change the namespace of mailboxes to new user when migrating, we don't create anything new in fact... Forget my previous comment :)

File: mailbox/api/src/main/java/org/apache/james/mailbox/quota/MaxQuotaManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/quota/MaxQuotaManager.java:57]: Please subscribe this on the blocking callwrapper scheduler

**Review by chibenwa — APPROVED** [review]
Just to be sure: prior to this current quota were not migrated, right?

**Comment by vttranlina** [comment]
> Just to be sure: prior to this current quota were not migrated, right?

yes

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by quantranhong1999** [comment]
rebase needed

**Comment by vttranlina** [comment]
rebase

**Pull request merged**